clippy_lints/src/outer_doc_comment_bang.rs

```diff
@@ -0,0 +1,99 @@
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use std::fmt::Write;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects the use of outer doc comments (`///`, `/**`) followed by a bang (`!`).
+    ///
+    /// ### Why is this bad?
+    /// Triple-slash comments (known as "outer doc comments") apply to items that follow it.
+    /// An outer doc comment followed by a bang (i.e. `///!`) has no specific meaning.
+    ///
+    /// The user most likely meant to use an inner doc comment (`//!`, `/*!`), which
+    /// applies to the parent item (i.e. the item that this comment is contained in,
+    /// usually a module or crate).
+    ///
+    /// ### Example
+    /// In this example, the doc comment is attached to the *function*, rather than the *module*.
+    /// ```rust
+    /// pub mod util {
+    ///     ///! This module contains utility functions.
+    ///
+    ///     pub fn dummy() {}
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust
+    /// pub mod util {
+    ///     //! This module contains utility functions.
+    ///
+    ///     pub fn dummy() {}
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub OUTER_DOC_COMMENT_BANG,
+    suspicious,
+    "outer doc comment followed by a bang symbol"
+}
+declare_lint_pass!(OuterDocCommentBang => [OUTER_DOC_COMMENT_BANG]);
+
+const WARNING: &str = "this is an outer doc comment and does not apply to the parent module or crate";
+const HELP: &str = "use an inner doc comment to document the parent module or crate";
+
+impl EarlyLintPass for OuterDocCommentBang {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        let mut lo_span = None;
+        let mut hi_span = None;
+        let mut sugg = String::new();
+
+        // multiline doc comments span across multiple attributes
+        // so keep pushing into a string
+        for attr in &item.attrs {
+            if_chain! {
+                if let AttrKind::DocComment(com_kind, sym) = attr.kind;
+                if let AttrStyle::Outer = attr.style;
+                if let Some(com) = sym.as_str().strip_prefix('!');
+                then {
+                    if !sugg.is_empty() {
+                        sugg.push('\n');
+                    }
+
+                    if lo_span.is_none() {
+                        lo_span = Some(attr.span);
+                    }
+
+                    hi_span = Some(attr.span);
+                    let _ = match com_kind {
+                        CommentKind::Line => write!(sugg, "//!{com}"),
+                        CommentKind::Block => write!(sugg, "/*!{com}*/")
+                    };
+                }
+            }
+        }
+
+        if let Some((lo_span, hi_span)) = lo_span.zip(hi_span) {
+            let span = lo_span.to(hi_span);
+            let indent = indent_of(cx, span);
+            let sugg = reindent_multiline(sugg.into(), true, indent);
```

Comment:
This suggestion takes all lines that are bad, but does not include the one that are legit.

e.g.

```rust
pub mod foo6 {
    ///! sugg1
    /// This is gone in sugg
    ///! sugg2
    
    pub fn bar() {}
}
```

produces

```stderr
error: this is an outer doc comment and does not apply to the parent module or crate
  --> $DIR/outer_doc_comment_bang.rs:36:5
   |
LL | /     ///! sugg1
LL | |     /// This is gone in sugg
LL | |     ///! sugg2
   | |______________^
   |
help: use an inner doc comment to document the parent module or crate
   |
LL ~     //! sugg1
LL +     //! sugg2
   |
```

Reply:
I think we can use `multispan` suggestion instead, or rather, just lint the lines individually if that's easier.

Reply:
> just lint the lines individually if that's easier.

This does seem a lot easier (and that's what I initially had). Gets rid of hacking the lo/hi spans together like this, but IME when someone writes crate-level or module-level documentation, it tends to be a long doc comment with lots of lines so it would report a lot of the same diagnostics potentially. I thought it'd be nicer if it could combine them all into one diagnostic since they are all logically part of the same doc comment.
I'm not sure if this really a problem and there seem to be quite a few corner cases I missed by doing it like this, so just linting each line individually as they are actually represented in the AST might be better.

Reply:
> I thought it'd be nicer if it could combine them all into one diagnostic since they are all logically part of the same doc comment

We can use [multispan_sugg](https://github.com/rust-lang/rust-clippy/blob/89160b6de2a62499a5c4a484024cd62061e3d2de/clippy_utils/src/diagnostics.rs#L222-L236) to do this. So it'll be one suggestion but with different spans